<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

In [2]:
RANDOM_STATE = 12345

In [3]:
df = pd.read_csv('datasets/insurance.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [4]:
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [5]:
df['Пол'] = df['Пол'].astype(float)
df['Члены семьи'] = df['Члены семьи'].astype(float)
df['Страховые выплаты'] = df['Страховые выплаты'].astype(float)
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0  1.0     41.0   49600.0          1.0                0.0
1  0.0     46.0   38000.0          1.0                1.0
2  0.0     29.0   21000.0          0.0                0.0
3  0.0     21.0   41700.0          2.0                0.0
4  1.0     28.0   26100.0          0.0                0.0

In [6]:
df.shape

(5000, 5)

In [7]:
df.values

array([[1.00e+00, 4.10e+01, 4.96e+04, 1.00e+00, 0.00e+00],
       [0.00e+00, 4.60e+01, 3.80e+04, 1.00e+00, 1.00e+00],
       [0.00e+00, 2.90e+01, 2.10e+04, 0.00e+00, 0.00e+00],
       ...,
       [0.00e+00, 2.00e+01, 3.39e+04, 2.00e+00, 0.00e+00],
       [1.00e+00, 2.20e+01, 3.27e+04, 3.00e+00, 0.00e+00],
       [1.00e+00, 2.80e+01, 4.06e+04, 1.00e+00, 0.00e+00]])

<div class="alert alert-block alert-info">
    
Пропуски в данных отсутствуют.
    
Ширина матрицы признаков - 5 колонок вместе с целевым признаком, значит размер обратимой матрицы для домножения будет 4 х 4.
    
Целевой признак - Страховые выплаты.
</div>

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:**

При умножении признаков на обратимую матрицу качество линейной регрессии не изменится, если результат умножения признаков на обратимую матрицу умножить на обратную матрицу от матрицы, на которую умножали.


**Обоснование:**

$$
w' = (X^T P^T X P)^{-1} X^T P^T y
$$

$$
w' = ((XP)^T XP)^{-1} (XP)^T y
$$

$$
w' = (P^T X^T XP)^{-1} P^T X^T y
$$

$$
w' = P^{-1} (X^T X)^{-1} (P^T)^{-1} P^T X^T y
$$

$$
w' = P^{-1} (X^T X)^{-1} X^T y
$$

$$
w' = P^{-1} w
$$

$$
w = P w'
$$


$$
a' = XP P^{−1} w
$$

$$
a' = Xw
$$


$$
a' = a
$$

<div class="alert alert-block alert-info">
При перемножении матриц порядок множителей имеет принципиальное значение, поэтому раскрываем скобки мы только на базе вот этих свойств:
$$
(AB)^T=B^T A^T
$$
$$
(AB)^{-1} = B^{-1} A^{-1}
$$
Произвольно менять порядок и переставлять матрицы как нам удобно мы не можем. Кроме того, мы производим домножение на случайную матрицу только справа, слева результаты линейной регрессии не будут совпадать
</div>

In [8]:
def check_score(f, t, custom_class=None, P=None):
    features_train, features_valid, target_train, target_valid = train_test_split(f, t, test_size=0.25, random_state=RANDOM_STATE)
    
    print('features_train.head', features_train.head())
    print()
    print('features_valid.head', features_valid.head())
    print()
    print('target_train.head', target_train.head())
    print()
    print('target_valid.head', target_valid.head())
    
    model = None
    if custom_class:
        model = custom_class()
    else:
        model = LinearRegression()
    
    if P is not None:
        model.fit(features_train, target_train, P)
    else:
        model.fit(features_train, target_train)
    target_predictions = model.predict(features_valid)
    
    score = r2_score(target_valid, target_predictions)
    return score
    

In [9]:
target = df['Страховые выплаты']
features = df.drop(['Страховые выплаты'], axis=1)
print('target.head', target.head())
print()
print('features.head', features.head())

target.head 0    0.0
1    1.0
2    0.0
3    0.0
4    0.0
Name: Страховые выплаты, dtype: float64

features.head    Пол  Возраст  Зарплата  Члены семьи
0  1.0     41.0   49600.0          1.0
1  0.0     46.0   38000.0          1.0
2  0.0     29.0   21000.0          0.0
3  0.0     21.0   41700.0          2.0
4  1.0     28.0   26100.0          0.0


Обучим линейную регрессию

In [10]:
check_score(features, target)

features_train.head       Пол  Возраст  Зарплата  Члены семьи
3369  1.0     43.0   36200.0          1.0
1441  1.0     34.0   57600.0          0.0
571   0.0     32.0   41100.0          1.0
225   0.0     36.0   45100.0          1.0
2558  0.0     33.0   50600.0          2.0

features_valid.head       Пол  Возраст  Зарплата  Члены семьи
3183  0.0     33.0   39000.0          4.0
1071  0.0     50.0   43100.0          2.0
2640  1.0     39.0   42100.0          0.0
2282  0.0     20.0   34800.0          0.0
1595  0.0     41.0   40000.0          4.0

target_train.head 3369    1.0
1441    0.0
571     0.0
225     0.0
2558    0.0
Name: Страховые выплаты, dtype: float64

target_valid.head 3183    0.0
1071    2.0
2640    0.0
2282    0.0
1595    0.0
Name: Страховые выплаты, dtype: float64


0.43522757127025635

Обучим линейную регрессию на закодированных данных

In [11]:
def inv_matrix():
    res = np.random.normal(size=(4,4))
    
    try:
        np.linalg.inv(res)
        return res
    except:
        inv_matrix()

In [12]:
random_matrix = inv_matrix()
random_matrix

array([[ 1.55225216, -1.10764209, -0.02076024,  0.58764478],
       [ 1.43902341,  0.30249596, -1.06500964, -0.58133143],
       [-0.0153801 , -0.77204302, -1.40605057, -0.83354945],
       [-0.32542082, -0.85580739, -1.86283929,  0.02482349]])

In [13]:
new_features = np.dot(features.values, random_matrix)
new_features = pd.DataFrame(new_features, columns=features.columns)
new_features

Пол       Возраст      Зарплата   Члены семьи
0    -702.626227 -38282.895096 -69785.657351 -41367.274955
1    -518.574189 -29324.575898 -53480.775007 -31701.595610
2    -281.250446 -16204.131117 -29557.947285 -17521.397110
3    -611.781569 -32189.553292 -58658.399721 -34771.170474
4    -359.575733 -20142.960676 -36727.760952 -21771.330340
...          ...           ...           ...           ...
4995 -509.427798 -27555.177678 -50229.551358 -29773.943080
4996 -757.315926 -40445.625392 -73715.123124 -43697.731746
4997 -493.255803 -26167.920202 -47690.140252 -28268.903415
4998 -470.694804 -25242.827031 -46006.893185 -27269.194267
4999 -582.912621 -31338.440329 -57117.357081 -33857.772575

[5000 rows x 4 columns]

In [14]:
check_score(new_features, target)

features_train.head              Пол       Возраст      Зарплата   Члены семьи
3369 -493.654825 -27936.913585 -50946.709710 -30198.874957
1441 -835.414780 -44460.500950 -81024.744018 -48031.626077
571  -586.398830 -31722.144215 -57824.621645 -34277.460272
225  -642.163141 -34809.106326 -63453.083970 -37613.983407
2558 -731.396189 -39057.106252 -71185.029925 -42196.736577

features_valid.head              Пол       Возраст      Зарплата   Члены семьи
3183 -553.637857 -30103.118791 -54878.568972 -32527.513283
1071 -591.582032 -33261.641143 -60657.755801 -35954.998319
2640 -589.828094 -32492.321602 -59236.285205 -35114.516223
2282 -506.447053 -26861.047309 -48951.860088 -29019.147569
1595 -557.505770 -30872.741847 -56293.139621 -33365.713387

target_train.head 3369    1.0
1441    0.0
571     0.0
225     0.0
2558    0.0
Name: Страховые выплаты, dtype: float64

target_valid.head 3183    0.0
1071    2.0
2640    0.0
2282    0.0
1595    0.0
Name: Страховые выплаты, dtype: float64


0.43522757127027567

Качество линейной регресии не изменилось.

## Алгоритм преобразования

**Алгоритм**

$$
a = XPP^{-1}w
$$

<div class="alert alert-block alert-info">
Мы производим домножение всегда справа.
</div>

**Обоснование**

...

In [15]:
class CustomLinearRegression:
    def fit(self, train_features, train_target, P=None):
        y = train_target
        w = None
        self.p = None
        
        if P is not None:
            new_features = train_features.values @ P
            X = np.concatenate((np.ones((new_features.shape[0], 1)), new_features), axis=1)
            w = np.linalg.inv(X.T @ X) @ X.T @ y
            self.p = P
            self.w = self.p @ w[1:]
        else:
            X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
            w = np.linalg.inv(X.T @ X) @ X.T @ y
            self.w = w[1:]
        
        self.w0 = w[0]

    def predict(self, test_features):
        return test_features @ self.w + self.w0

## Проверка алгоритма

In [16]:
check_score(features, target, CustomLinearRegression)

features_train.head       Пол  Возраст  Зарплата  Члены семьи
3369  1.0     43.0   36200.0          1.0
1441  1.0     34.0   57600.0          0.0
571   0.0     32.0   41100.0          1.0
225   0.0     36.0   45100.0          1.0
2558  0.0     33.0   50600.0          2.0

features_valid.head       Пол  Возраст  Зарплата  Члены семьи
3183  0.0     33.0   39000.0          4.0
1071  0.0     50.0   43100.0          2.0
2640  1.0     39.0   42100.0          0.0
2282  0.0     20.0   34800.0          0.0
1595  0.0     41.0   40000.0          4.0

target_train.head 3369    1.0
1441    0.0
571     0.0
225     0.0
2558    0.0
Name: Страховые выплаты, dtype: float64

target_valid.head 3183    0.0
1071    2.0
2640    0.0
2282    0.0
1595    0.0
Name: Страховые выплаты, dtype: float64


0.43522757127026657

In [17]:
P = np.random.normal(size=(features.shape[1], features.shape[1]))
print(features.shape)
print('P shape', P.shape)
print()
check_score(features, target, CustomLinearRegression, P)

(5000, 4)
P shape (4, 4)

features_train.head       Пол  Возраст  Зарплата  Члены семьи
3369  1.0     43.0   36200.0          1.0
1441  1.0     34.0   57600.0          0.0
571   0.0     32.0   41100.0          1.0
225   0.0     36.0   45100.0          1.0
2558  0.0     33.0   50600.0          2.0

features_valid.head       Пол  Возраст  Зарплата  Члены семьи
3183  0.0     33.0   39000.0          4.0
1071  0.0     50.0   43100.0          2.0
2640  1.0     39.0   42100.0          0.0
2282  0.0     20.0   34800.0          0.0
1595  0.0     41.0   40000.0          4.0

target_train.head 3369    1.0
1441    0.0
571     0.0
225     0.0
2558    0.0
Name: Страховые выплаты, dtype: float64

target_valid.head 3183    0.0
1071    2.0
2640    0.0
2282    0.0
1595    0.0
Name: Страховые выплаты, dtype: float64


0.4352275712650374

Метрика R2 показала результат 0.435

<div class="alert alert-block alert-success">
Дополнительные материалы:

Обзоры по применению криптографии в машинном обучении: https://arxiv.org/pdf/1902.04109.pdf, https://www.researchgate.net/publication/342746786_Some_Applications_of_Machine_Learning_in_Cryptography

Небольшой обзор на Шифр Хилла: https://habr.com/ru/post/595281/
   
Примеры использования линейной алгебры в машинном обучении:
https://machinelearningmastery.com/examples-of-linear-algebra-in-machine-learning/
    
Для чего нужна математика в машинном обучении?
https://proglib.io/p/dlya-chego-nuzhen-matematicheskiy-analiz-v-mashinnom-obuchenii-2021-02-15

Большая статья по векторам:
https://neptune.ai/blog/understanding-vectors-from-a-machine-learning-perspective  

Можно ещё дополнительно посмотреть лекции от МФТИ https://www.youtube.com/playlist?list=PLk4h7dmY2eYHHTyfLyrl7HmP-H3mMAW08
    
Классная серия роликов по линейной алгебре на английском (можно включить автоматические русские субтитры): https://www.youtube.com/playlist?list=PLRDl2inPrWQW1QSWhBU0ki-jq_uElkh2a
    
Удачи в будущих проектах и работе =)
</div>
